### PR TITLE
Properly pass `null` exception to error handlers

### LINKED ISSUE #3183: `throw undefined` turns into a `TypeError: Cannot read property 'message' of undefined`
### Basic info:

- **Node.js version:** 12.16.1
- **jsdom version:** 16.5.1

### Minimal reproduction case

```js
const { JSDOM } = require("jsdom");

const dom = new JSDOM(`<div id="test"></div>`);
const test = dom.window.document.getElementById("test");
test.onclick = () => { throw undefined; };
test.click();
```

result:
```
Uncaught TypeError: Cannot read property 'message' of undefined
    at reportException (E:\Code\XS.AMC.UX\node_modules\jsdom\lib\jsdom\living\helpers\runtime-script-errors.js:58:84)
    at innerInvokeEventListeners (E:\Code\XS.AMC.UX\node_modules\jsdom\lib\jsdom\living\events\EventTarget-impl.js:341:9)
    at invokeEventListeners (E:\Code\XS.AMC.UX\node_modules\jsdom\lib\jsdom\living\events\EventTarget-impl.js:274:3)
    at HTMLDivElementImpl._dispatch (E:\Code\XS.AMC.UX\node_modules\jsdom\lib\jsdom\living\events\EventTarget-impl.js:221:9)
    at fireAnEvent (E:\Code\XS.AMC.UX\node_modules\jsdom\lib\jsdom\living\helpers\events.js:18:36)
    at HTMLDivElementImpl.click (E:\Code\XS.AMC.UX\node_modules\jsdom\lib\jsdom\living\nodes\HTMLElement-impl.js:79:5)
    at HTMLDivElement.click (E:\Code\XS.AMC.UX\node_modules\jsdom\lib\jsdom\living\generated\HTMLElement.js:110:34)
```

### How does similar code behave in browsers?

image

### Why is this important?

We're trying to test a React component that throws `undefined` and is caught by an error boundary. But instead of the error boundary seeing `undefined`, it sees an `Error` with a message of `Cannot read property 'message' of undefined`.

### PR DESCRIPTION
**Pull request by mbest**
Check for `null` or `undefined` error values before trying to access properties.

Fixes #3183

### CODE REVIEW

**Review by domenic — CHANGES_REQUESTED** [review]
Thanks for working on this!

Per CONTRIBUTING.md, no new tests should be added to `to-port-to-wpts/` (as that increases our porting burden). Instead, the test for the `onerror` handler should be added to the appropriate location in `web-platform-tests/to-upstream`, and the test for the virtual console behavior should be added to `api/jsdom-errors.js`

**Comment by mbest** [comment]
Can you tell me the *appropriate* location?

> Per CONTRIBUTING.md, no new tests should be added to `to-port-to-wpts/` (as that increases our porting burden). Instead, the test for the `onerror` handler should be added to the appropriate location in `web-platform-tests/to-upstream`, and the test for the virtual console behavior should be added to `api/jsdom-errors.js`

**Review by domenic — COMMENTED** [review]
I've told GitHub to run the CI. I think it will fail on the lint in at least some cases.

Thanks for moving the tests. I think a better location for web platform tests would be `html/webappapis/scripting/processing-model-2` (you'll need to create a couple folders), instead of `dom/events`. That's where [existing tests for uncaught exceptions ](https://github.com/web-platform-tests/wpt/tree/master/html/webappapis/scripting/processing-model-2) live in upstream.

**Comment by TimothyGu** [comment]
@mbest As an example, https://github.com/web-platform-tests/wpt/blob/master/html/webappapis/scripting/processing-model-2/window-onerror-parse-error.html is a test that tests the `window.onerror` event handler.

File: test/web-platform-tests/to-upstream/dom/events/Event-dispatch-throwing.html [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+<script src="/resources/testharnessreport.js"></script>
+<div id="log"></div>
+<script>
+  "use strict";
+
+  setup({ allow_uncaught_exception: true });
+
+  test(() => {
+    let errorEvents = 0;
+    document.defaultView.addEventListener("error", event => {
```
Comment (domenic) [inline test/web-platform-tests/to-upstream/dom/events/Event-dispatch-throwing.html:14]: In addition to moving this file into the correct location, it's best to update this to just `window.addEventListener`

File: test/web-platform-tests/to-upstream/dom/events/Event-dispatch-throwing.html [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    element.addEventListener("click", () => {
+      throw null; // eslint-disable-line no-throw-literal
+    });
+
+    element.dispatchEvent(new Event("click"));
+
+    assert_equals(errorEvents, 1);
+  }, "Throwing null in event listener");
+</script>
```
Comment (domenic) [inline test/web-platform-tests/to-upstream/dom/events/Event-dispatch-throwing.html:30]: Newline at end of file please. (Not sure why lint didn't catch this...)

**Review by domenic — CHANGES_REQUESTED** [review]
Almost there :)

File: test/web-platform-tests/to-upstream/html/webappapis/scripting/processing-model-2/window-onerror-event-throw-null.html [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+<div id="log"></div>
+<script>
+  "use strict";
+
+  setup({ allow_uncaught_exception: true });
+
+  test(() => {
+    let errorEvents = 0;
+    window.addEventListener("error", event => {
+      assert_equals(event.message, null);
```
Comment (TimothyGu) [inline test/web-platform-tests/to-upstream/html/webappapis/scripting/processing-model-2/window-onerror-event-throw-null.html:15]: It looks like we technically cannot use `null` as `event.message`, as it is [defined](https://html.spec.whatwg.org/multipage/webappapis.html#runtime-script-errors:dom-errorevent-message-2) to always be a string.

I suggest this logic, which emulates the Firefox behavior:
```js
let message;
if (error === undefined || error === null || typeof error === "boolean" ||
    typeof error === "number" || typeof error === "bigint" || typeof error === "symbol") {
  message = "uncaught exception: " + String(error);
} else if (error.message) {
  message = String(error.message);
} else {
  message = "uncaught exception";
}
```

**Review by TimothyGu — APPROVED** [review]
Sorry, one last thing. Otherwise LGTM.

File: test/web-platform-tests/to-upstream/html/webappapis/scripting/processing-model-2/window-onerror-event-throw-null.html (hunk 3 above)
Comment (mbest) [reply test/web-platform-tests/to-upstream/html/webappapis/scripting/processing-model-2/window-onerror-event-throw-null.html]: Updated. The code I added matches the code for unhandled exceptions (line 65).

**Review by TimothyGu — APPROVED** [review]

**Comment by mbest** [comment]
@domenic Anything else?

**Review by domenic — APPROVED** [review]

**Pull request merged**